cpp/src/reductions/scan/ewm.cu

```diff
@@ -0,0 +1,377 @@
+                        return {beta, input};
+                      });
+
+    pair_beta_adjust(input, pairs, nullcnt, stream);
+    compute_recurrence(pairs, stream);
+
+    // copy the second elements to the output for now
+    thrust::transform(rmm::exec_policy(stream),
+                      pairs.begin(),
+                      pairs.end(),
+                      output.begin(),
+                      [=] __device__(pair_type pair) -> T { return pair.second; });
+
+    // Denominator
+    // Fill with pairs
+    thrust::fill(rmm::exec_policy(stream), pairs.begin(), pairs.end(), pair_type(beta, 1.0));
+
+    pair_beta_adjust(input, pairs, nullcnt, stream);
+    compute_recurrence(pairs, stream);
+
+    thrust::transform(
+      rmm::exec_policy(stream),
+      pairs.begin(),
+      pairs.end(),
+      output.begin(),
+      output.begin(),
+      [=] __device__(pair_type pair, T numerator) -> T { return numerator / pair.second; });
+
+  } else {
+    // Numerator
+    // Fill with pairs
+    thrust::transform(rmm::exec_policy(stream),
+                      input.begin<T>(),
+                      input.end<T>(),
+                      pairs.begin(),
+                      [=] __device__(T input) -> pair_type {
+                        return {beta, input};
+                      });
+    compute_recurrence(pairs, stream);
+
+    // copy the second elements to the output for now
+    thrust::transform(rmm::exec_policy(stream),
+                      pairs.begin(),
+                      pairs.end(),
+                      output.begin(),
+                      [=] __device__(pair_type pair) -> T { return pair.second; });
+
+    // Denominator
+    // Fill with pairs
+    thrust::fill(rmm::exec_policy(stream), pairs.begin(), pairs.end(), pair_type(beta, 1.0));
+    compute_recurrence(pairs, stream);
+
+    thrust::transform(
+      rmm::exec_policy(stream),
+      pairs.begin(),
+      pairs.end(),
+      output.begin(),
+      output.begin(),
+      [=] __device__(pair_type pair, T numerator) -> T { return numerator / pair.second; });
+  }
+  return output;
+}
+
+template <typename T>
+rmm::device_uvector<T> compute_ewma_noadjust(column_view const& input,
+                                             T beta,
+                                             rmm::cuda_stream_view stream,
+                                             rmm::mr::device_memory_resource* mr)
+{
+  using pair_type = thrust::pair<T, T>;
+  rmm::device_uvector<T> output(input.size(), stream, mr);
+  rmm::device_uvector<pair_type> pairs(input.size(), stream);
+
+  thrust::transform(rmm::exec_policy(stream),
+                    input.begin<T>(),
+                    input.end<T>(),
+                    pairs.begin(),
+                    [=] __device__(T input) -> pair_type {
+                      return {beta, (1.0 - beta) * input};
+                    });
+
+  // TODO: the first pair is WRONG using the above. Reset just that pair
```

Comment:
Still relevant?

Reply:
I was wondering here if there was a way of just setting one element (the first one) to be `{beta, input}` rather than having to use a whole transform?

Reply:
A slightly nicer (but still not amazing) solution would be to add a counting iterator to the above transform and condition the return value on whether that input value is equal to 0. It adds some overhead to the evaluation, but avoiding the extra kernel launch here should make that approach always win since you'll only see divergence for warp 0.

I can't think of anything cleaner than that unfortunately.